Proposition de fiche duckdb (#508)
InseeFrLab/utilitR: 03_Fiches_thematiques/Fiche_duckdb.qmd

- Il est essentiel de travailler avec la dernière version d'`arrow`, de `duckdb` et de `R` car les *packages* `arrow` et `duckdb` sont en cours de développement.

- Si les données sont très volumineuses (plus de 5 Go en CSV, plus de 1 Go en Parquet ou plus de 5 millions d'observations), il est essentiel de manipuler des requêtes `duckdb`, plutôt que des `tibbles` pour pouvoir réaliser les traitements sur une machine standard. Les éléments présentés dans cette fiche concernent préférentiellement ce cas de figure.
:::

::: {.callout-note}
Apprendre à utiliser `duckdb` n'est pas difficile, car la syntaxe utilisée est quasiment identique à celle du `tidyverse`. Toutefois, une bonne compréhension du fonctionnement de `R` et de `duckdb` est nécessaire pour bien utiliser `duckdb` sur des données volumineuses. Voici quelques conseils pour bien démarrer:

> -   Il est indispensable de lire la fiche [Manipuler des données avec le `tidyverse`](#tidyverse) avant de lire la présente fiche.

Reviewer: ```suggestion
-   Il est indispensable de lire la fiche [Manipuler des données avec le `tidyverse`](#tidyverse) avant de lire la présente fiche.
-   Il est recommandé de lire la fiche [Se connecter à une base de données](#bdd) avant de lire la présente fiche.
```

-   Il est complètement normal de rencontrer des erreurs difficiles à comprendre lorsqu'on commence à utiliser `duckdb`, il ne faut donc pas se décourager.
-   Il ne faut pas hésiter à demander de l'aide à des collègues, ou à poser des questions sur les salons Tchap adaptés (le salon Langage `R` par exemple).
:::

## Présentation du package `duckdb`  et du projet associé

### Qu'est-ce que `duckdb`?  {#sec-presentation}

[`DuckDB`](https://duckdb.org/) est un projet *open-source* (license MIT) qui propose un moteur SQL optimisé pour les requêtes *online analytical processing* (OLAP): 

- un moteur SQL rapide, capable d'utiliser des données au format `parquet` sans les charger en mémoire,

> - un dialect SQL enrichi pour l'analyse de données,

Author: Tous les moteurs SQL (les logiciels) implémentent des versions légèrement différentes de SQL. 
Ca se voit par exemple sur la clause LIMIT pour sélectionner les premières lignes. C'est TOP en Microsoft SQL, FETCH FIRST en oracle ...
En duckdb, il y a des fonctions particulières, et aussi des syntaxes. On peut faire SELECT EXCEPT pour éliminer des colonnes.

> - un dialect SQL enrichi pour l'analyse de données,

Reviewer 2: On a si je comprends bien la documentation de duckdb une syntaxe normalement extrêmement proche de PostgreSQL : https://duckdb.org/why_duckdb#feature-rich (en particulier : "SQL Parser: We use the PostgreSQL parser that was [repackaged as a stand-alone library](https://github.com/lfittl/libpg_query).").

- une installation et un déploiement facile,
- un moteur portable, utilisable sous Windows, MacOS, Linux, et interfacé avec de nombreux langages de programmation (R, Python, Javascript, etc.).

> `DuckDB` est bien un serveur SQL. Il dispose de sa propre mémoire, et on peut lui envoyer des requêtes SQL sans passer par R. 

Author: Personnellement, je n'entrerais pas dans ces arcanes, qui concernent plutôt des gens de type DSI. On peut préciser qu'il suffit d'installer le package

> `DuckDB` est bien un serveur SQL. Il dispose de sa propre mémoire, et on peut lui envoyer des requêtes SQL sans passer par R. 

Reviewer: OK, parfait

-   __*Évaluation différée*__: `duckdb` construit des requêtes SQL, qui sont exécutées uniquement quand nécessaire, après optimisation des étapes intermédiaires, et peuvent être exécutées partiellement. La @sec-lazy présente en détail cette notion.
-   __*Traduction en SQL*__: `duckdb` construit des requêtes SQL, et toutes les fonctions de `dplyr` ne sont pas traduites. Il y moins de limitations que `arrow` mais il faut parfois utiliser une fonction SQL directement ou contourner le problème. La @sec-sql donne quelques trucs et astuces dans ce cas.
-   __Interopérabilité__: `duckdb` est conçu pour être interopérable entre plusieurs langages de programmation tels que `R`, Python, Java, C++, etc. Cela signifie que les données peuvent être échangées entre ces langages sans avoir besoin de convertir les données, d'où des gains importants de temps et de performance.

# Utilisation de `duckdb`

> Dans cette section, on présente l'utilisation basique de `duckdb`. C'est très facile car il suffit de manipuler la syntaxe `dplyr`. En particulier, il n'est pas nécessaire d'avoir des notions de SQL. 

Reviewer: ```suggestion
Dans cette section, on présente l'utilisation basique de `duckdb`. C'est très facile: il n'est pas nécessaire de connaître le langage de SQL car il est possible d'utiliser `duckdb` avec la syntaxe `tidyverse`.
```

## Charger le *package* `duckdb`

Pour utiliser `duckdb`, il faut commencer par charger le *package*. Il est utile de charger le package `dplyr` pour bénéficier de la syntaxe correspondante. 

```{r}
#| output: false
library(duckdb)
library(dplyr)
```

Le moteur `duckdb` fonctionnant "en dehors" de R, il détecte le nombre de processeurs et effectue les opérations en parallèle si possible. 

## Connexion à une base de données

> **`duckdb` est une base de données distante et s'utilise comme tel: il faut ouvrir une connexion et "charger" les données.** 

Reviewer: ```suggestion
**`duckdb` est une base de données distante et s'utilise comme telle: il faut ouvrir une connexion, puis "charger" les données dans la base de données pour les manipuler.** 
```

Comme beaucoup d'autres bases de données (distantes ou locales), on ouvre une connexion au moteur `duckdb` avec une base de données en mémoire vive de la façon suivante :
```{r}
con <- DBI::dbConnect(duckdb::duckdb()); 

> ```

Reviewer: Le point-virgule est-il important?

Author: Le point-virgule n'est pas important, j'ai effectué la modif.

Reviewer: OK, parfait

> 

Reviewer: ```suggestion

Concrètement, cette commande crée une nouvelle base de données `duckdb` dans la mémoire vive. Cette base de données ne contient aucune donnée lorsqu'elle est créée.

```

À la fin du traitement ou du programme, on ferme la connexion avec le code ci-dessous. L'option `shutdown` est importante : elle permet de fermer le moteur et de récupérer la mémoire. Autrement, il arrive souvent d'avoir d'avoir des connexions à moitié ouvertes, et de devoir relancer la session R.
```{r}
DBI::dbDisconnect(con, shutdown = TRUE)
```

Pour la suite, on supposera que la connexion est ouverte.
```{r}
con <- DBI::dbConnect(duckdb::duckdb()); 
```

> ## Chargement des données et écriture au format parquet

Reviewer: Je pense qu'il faut bien séparer les deux sujets: chargement des données et écriture au format Parquet. Il me semble qu'un plan logique serait: 1/ charger le _package_; 2/ ouvrir la connexion; 3/ charger des données; 4/ manipuler des données; 5/ écrire en Parquet.

Author: Oui mais. Je pense qu'il faut insister sur "lire directement depuis les fichiers parquet", et du coup j'ai besoin d'écrire et lire un fichier avant de manipuler. J'essaye de proposer une modification.

> ### Déclaration de données provenant de R

Reviewer: ```suggestion
### Déclaration de données provenant de `R`
```

# Manipuler des données avec `duckdb` {#duckdb}

## Tâches concernées et recommandations

L'utilisateur souhaite manipuler des données structurées sous forme de `data.frame` par le biais de l'écosystème `duckdb` (sélectionner des variables, sélectionner des observations, créer des variables, joindre des tables).

::: {.callout-important}
Tâches concernées et recommandations

> - Pour des tables de données de taille petite et moyenne (inférieure à 1 Go ou moins d'un million d'observations), il est recommandé d'utiliser les *packages* `tibble`, `dplyr` et `tidyr` qui sont présentés dans la fiche [Manipuler des données avec le `tidyverse`](#tidyverse);

Reviewer 2: De manière générale, je suis d'accord avec les recommandations qui sont formulées ici. En revanche, je serais assez favorable à les positionner ailleurs que dans la seule fiche sur `duckdb` (_cf_. la discussion sur le sujet dans l'issue #499).

Author: D'accord pour les positionner aussi ailleurs. C'est une discussion qui dépasse la seule fiche duckdb.

Reviewer 3: Oui on harmonisera ces recommandations ultérieurement

- Pour des tables de données de grande taille (plus de 1 Go en CSV, plus de 200 Mo en Parquet, ou plus d'un million d'observations), il est recommandé d'utiliser soit le *package* `data.table` qui fait l'objet de la fiche [Manipuler des données avec `data.table`](#datatable) lorsque les données tiennent en mémoire vive, soit le *package* `arrow` [Manipuler des données avec `arrow`](#arrow). Nous présentons ici les fonctionnalités du package `duckdb` en complément ou en remplacement de `arrow`.

- Il est essentiel de travailler avec la dernière version d'`arrow`, de `duckdb` et de `R` car les *packages* `arrow` et `duckdb` sont en cours de développement.

> - Si les données sont très volumineuses (plus de 5 Go en CSV, plus de 1 Go en Parquet ou plus de 5 millions d'observations), il est essentiel de manipuler des requêtes `duckdb`, plutôt que des `tibbles` pour pouvoir réaliser les traitements sur une machine standard. Les éléments présentés dans cette fiche concernent préférentiellement ce cas de figure.

Reviewer 2: Je modérerais bien le propos. Au-delà d'un certain volume, il peut être optimal de basculer sur des solutions plus orientées big-data. Il ne faudrait pas laisser penser que duckdb est la solution "one-size-fits-all" sur de gros volumes de données.

Author: J'ai reformulé un peu. Dis-moi ce que tu en penses

- un dialecte SQL enrichi avec des fonctions qui facilitent l'analyse de données,
- une installation et un déploiement facile,
- un moteur portable, utilisable sous Windows, MacOS, Linux, et interfacé avec de nombreux langages de programmation (R, Python, Javascript, etc.).

Un point important à comprendre est que `DuckDB` est un système de gestion de base de données (SGBD), similaire par exemple à un serveur PostgreSQL: il dispose donc de sa propre mémoire, et on peut lui envoyer directement des requêtes SQL, sans passer par `R`.
Un point important à retenir est donc que **`DuckDB` n'est pas un outil spécifique à `R`**, et il faut bien distinguer le projet `DuckDB` du *package* `R` `duckdb`. Ce *package* propose simplement une interface avec `R` parmi les autres interfaces existantes : Python, Java, Javascript, Julia, etc.

Toutefois, `DuckDB` est très facile à utiliser avec `R`, ce qui permet de bénéficier des optimisations inhérentes au langage SQL, à la fois en terme d'utilisation de la mémoire et de rapidité de calcul. C'est de plus un bon intermédiaire avant de passer à des infrastructures avancées telles que spark ou oracle.

> ### A quoi sert le *package* `duckdb`?

Reviewer 2: ```suggestion
### À quoi sert le *package* `duckdb`?
```

Du point de vue d'un statisticien utilisant `R`, le *package* `duckdb` permet de faire trois choses:

-   Importer des données (exemples: fichiers CSV, fichiers Parquet);
-   Manipuler des données avec la syntaxe `dplyr`, ou avec le langage SQL;
-   Écrire des données en format Parquet.

### Quels sont les avantages de `duckdb`?

-   **Disponibilité immédiate**: on peut pré-visualiser les données ou le résultat d'un calcul sans l'exécuter totalement, sans attendre le chargement des données;
-   **Performances élevées**: `duckdb` est très rapide pour la manipulation de données tabulaires (nettement plus performant que `dplyr` par exemple);

> -   **Ne pas charger les données en mémoire**: `duckdb` permet de travailler directement sur des fichiers du disque dur;

Reviewer 2: ```suggestion
-   **Ne pas nécessairement charger les données en mémoire**: `duckdb` permet également de travailler directement sur des fichiers du disque dur;
```

-   **Optimisations automatiques**: `duckdb` sélectionne automatiquement les colonnes nécessaires, et ne lit que les lignes nécessaires. Cela permet d'accélérer les calculs et de réduire considérablement les besoins en mémoire, même lorsque les données sont volumineuses;
-   **Facilité d'apprentissage** grâce aux approches `dplyr` et SQL: `duckdb` peut être utilisé avec les verbes de `dplyr` (`select`, `mutate`, etc.) et/ou avec le langage SQL. Par conséquent, il n'est pas nécessaire d'apprendre une nouvelle syntaxe pour utiliser `duckdb`, on peut s'appuyer sur la ou les approches que l'on maîtrise déjà. 

### Quelles sont les points d'attention à l'usage ?

-   __Représentation des données en mémoire__ : `duckdb` est un moteur SQL. Les lignes n'ont pas d'ordre pré-défini, et le résultat d'un traitement peut être dans un ordre aléatoire.

> -   __Traitement de données volumineuses__: `duckdb` traite des données qui peuvent être en mémoire vive ou sur le disque dur. En mémoire vive, les échanges sont efficaces avec `arrow` mais il faudra convertir les données pour les utiliser au format `tibble` (fonction `collect()`) ou avec `data.table`. Avec des données sur le disque dur, `duckdb` est capable de faire les traitements sur des données plus volumineuses que la mémoire vive (RAM). C'est un avantage majeur en comparaison aux autres approches possibles en `R` (`data.table` et `dplyr` par exemple).

Reviewer 2: ```suggestion
-   __Traitement de données volumineuses__: `duckdb` traite des données qui peuvent être en mémoire vive ou sur le disque dur. En mémoire vive, les échanges sont efficaces avec `arrow` mais il faudra convertir les données pour les utiliser au format `tibble` (fonction `collect()`) ou avec `data.table`. Avec des données sur le disque dur, `duckdb` est capable de faire les traitements sur des données plus volumineuses que ce que la mémoire vive (RAM) peut contenir, en contrepartie de temps de calcul plus longs. C'est un avantage majeur en comparaison aux autres approches possibles en `R` (`data.table` et `dplyr` par exemple).
```

Author: On paye la lecture des données, mais on ne pourrait pas faire autrement. Ce n'est pas vraiment comparable.
On peut reformuler en disant : C'est un avantage majeur. On obtient le résultat, mais il faut ajouter le temps de lecture des données.

-   __*Traduction en SQL*__: `duckdb` traduit automatiquement les instructions `dplyr` en requêtes SQL (de la même façon qu'`arrow` traduit ces instructions en code C++). Il arrive toutefois que certaines fonctions de `dplyr` n'aient pas d'équivalent direct en `duckdb` et ne puissent être traduites automatiquement. Dans ce cas (qui est heureusement moins fréquent qu'avec `arrow`), il faut parfois utiliser une fonction SQL directement ou trouver une solution pour contourner le problème. La @sec-sql donne quelques trucs et astuces dans ce cas.
-   __Interopérabilité__: `duckdb` est conçu pour être interopérable entre plusieurs langages de programmation tels que `R`, Python, Java, C++, etc. Cela signifie que les données peuvent être échangées entre ces langages sans avoir besoin de convertir les données, d'où des gains importants de temps et de performance.

# Installation de `duckdb`

Il suffit d'installer le package `duckdb` à partir du CRAN.

> 

Reviewer 2: ```suggestion
```{r}
install.packages("duckdb", repos="https://cloud.r-project.org")
```

```{r}
con %>% duckdb::duckdb_register(
  name = "bpe_ens_2018_duckdb", 
  df = doremifasolData::bpe_ens_2018 |> as_tibble())
```

Le code ci-dessous permet de vérifier que le chargement des données a bien fonctionné. La fonction `tbl` permet d'accéder à un objet de la base de données par le nom (de la table), ou par du code SQL (utilisation avancée). Par défaut, `duckdb` affiche les 10 premières lignes du résultat, sans effectuer tout le calcul. C'est super pratique !
```{r}
con %>% tbl("bpe_ens_2018_duckdb")
```

### Écriture au format parquet 

**Pour écrire une table (ou n'importe quelle requête) sur le disque au format parquet, on suggère d'utiliser la librairie `arrow`.**
```{r}
con %>% tbl("bpe_ens_2018_duckdb") %>% 
  arrow::to_arrow() %>% arrow::write_dataset("bpe_ens_2018_dataset")
list.files("bpe_ens_2018_dataset") # liste des fichiers du répertoire bpe_ens_2018_dataset/
```

> Une autre possibilité est de passer par la commande SQL `COPY ... TO`. Dans mon expérience, passer par `arrow` permet d'exécuter des calculs plus complexes sans faire déborder la mémoire (`arrow` sauvegarde au fur et à mesure), et la syntaxe est plus facile à manipuler, notamment quand on souhaite ajouter des options.

Reviewer 2: ```suggestion
Une autre possibilité est de passer par la commande SQL `COPY ... TO`. D'expérience, passer par `arrow` permet d'exécuter des calculs plus complexes sans faire déborder la mémoire (`arrow` sauvegarde au fur et à mesure), et la syntaxe est plus facile à manipuler, notamment quand on souhaite ajouter des options.
```

Reviewer 4: D'après mes tests, le `COPY ... TO` était un peu plus efficace que le `arrow::write_dataset`, il sauvegarde aussi au fur et à mesure :
https://nbc.github.io/notes-en-vrac/cas-pratique-traitement-duckdb.html#la-solution-arrowwrite_dataset

Author: Je n'ai jamais réussi à faire marcher `COPY .. TO` avec une sauvegarde au fur et à mesure. 
Dans mes benchmarks (sur fichier moyen), il me faut 500Mo de RAM en passant par `COPY ... TO` avec `per_thread_output`, mais seulement 108Mo avec `write_dataset`

- lors d'un appel à `print(n=nb_lignes)`: le calcul est exécuté partiellement, et seules les `nb_lignes` demandées sont affichées.

En utilisant `print()`, on peut prévisualiser le résultat d'une requête `duckdb`. Il ne faut pas hésiter à s'en servir pour explorer les données et pour construire son calcul itérativement, et ajuster en fonction des résultats.

## Erreurs courantes 

**On a éliminé des colonnes nécessaires**

```{r error=TRUE}
con %>% tbl("bpe_ens_2018_duckdb") |> select(DEP) |>
      mutate(NB_EQUIP_TOTAL_DEP  = sum(NB_EQUIP))
```

**Convertir les types**

Dans cet exemple, on veut multiplier un nombre par une indicatrice.
```{r error=TRUE}
con %>% tbl("bpe_ens_2018_duckdb") %>%
  summarise(nb_boulangeries  = sum(NB_EQUIP * (TYPEQU == "B203")), .by = DEP)
```

Avec `duckdb`, il faut transformer explicitement un booléen en nombre (entier ou flottant).

```{r}
con %>% tbl("bpe_ens_2018_duckdb") %>%
  summarise(nb_boulangeries  = sum(NB_EQUIP * as.integer(TYPEQU == "B203")), .by = DEP)
```

# Notions avancées / bien utiliser `duckdb`

## L'évaluation différée avec `duckdb` (_lazy evaluation_) {#sec-lazy}

> Quand on manipule des objets `duckdb`, on construit des requêtes SQL. Le package `duckdb` se contente de traduire le code `R` en `SQL` sans l'exécuter. On rappelle qu'il faut utiliser `show_query()` pour visualiser la requête. La fonction `print()` pré-visualise le résultat.

Reviewer 2: ```suggestion
Quand on manipule des objets `duckdb`, on construit des requêtes SQL. Le package `duckdb` se contente de traduire le code `R` en `SQL` sans l'exécuter. On rappelle qu'il faut utiliser `show_query()` pour visualiser la requête. La fonction `print()` permet de pré-visualiser le résultat.
```

**Avancé :** utilisation sur plusieurs variables avec `mutate_at`.

```{r}
liste_variables <- c("date_naissance","date_deces")
con %>% tbl("dates_duckdb") %>% 
  mutate_at(liste_variables, ~ strptime(.,"%d/%m/%Y"))
```

## Exécuter du code SQL directement

`DuckDB` étant un serveur SQL à part entière, on peut interagir avec `DuckDB` directement avec des requêtes SQL. Par exemple,

```{r}
DBI::dbGetQuery(con, "SELECT * FROM bpe_ens_2018_duckdb") |> head()
```

Par exemple, on peut créer des vues ou des tables explicitement. La fonction `dbExecute` retourne le nombre de lignes modifiées, tandis que la fonction `dbGetQuery` retourne le résultat sous la forme d'un `tibble`.
```{r}
DBI::dbExecute(con, "CREATE TABLE bpe_ens_2018_table AS SELECT REG, SUM(NB_EQUIP) AS NB_EQUIP_TOT FROM bpe_ens_2018_duckdb GROUP BY REG") # Éviter
DBI::dbExecute(con, "CREATE VIEW bpe_ens_2018_view AS SELECT REG, SUM(NB_EQUIP) AS NB_EQUIP_TOT FROM bpe_ens_2018_duckdb GROUP BY REG")   # n'utilise pas de mémoire
```

Et interroger les objets créés dans la base SQL via `dplyr` :
```{r}
con %>% tbl("bpe_ens_2018_view")
```

> Il est déconseillé de faire CREATE TABLE car cela charge les données dans la mémoire. Les fonctions `read_parquet()` en SQL et `duckdb_register` du package utilisent `CREATE VIEW` implicitement.

Reviewer 2: ```suggestion
Il est déconseillé de faire `CREATE TABLE` car cela charge les données dans la mémoire. Les fonctions `read_parquet()` en SQL et `duckdb_register` du package utilisent `CREATE VIEW` implicitement.
```

## Optimisations

Les opérations difficiles en SQL, longues, nécessitant beaucoup de mémoire, sont les fonctions "fenêtre" : jointures, `GROUP BY` avec beaucoup de petits groupes, dédoublonnage, etc. On propose ici quelques techniques pour faire passer ces calculs difficiles.

**Utilisation de la mémoire vive**

Comme expliqué plus haut, les objets manipulés dans cette fiche sont des requêtes SQL, et ne nécessitent pas de mémoire vive. Les données déclarées par `read_parquet` sont stockées sur le disque dur, lues à la demande, et "oubliées" à la fin du calcul. On retourne le _résultat_ du calcul.

Pour les opérations compliquées, il peut être nécessaire de charger les données en mémoire pour effectuer le calcul. En cas de débordement, `duckdb` renvoit un message du type :
```
Error: rapi_execute: Failed to run query
Error: Out of Memory Error: could not allocate block of size 262KB (99.7MB/100.0MB used)
Database is launched in in-memory mode and no temporary directory is specified.
Unused blocks cannot be offloaded to disk.

Launch the database with a persistent storage back-end
Or set PRAGMA temp_directory='/path/to/tmp.tmp'
```

> Pour éviter cela, il est possible d'adosser un fichier sur le disque dur à la base de données en mémoire au moment de la création de la connexion. Cela ralentit considérablement les calculs, et ne permet pas toujours d'obtenir un résultat. Dans mon expérience, il est préférable de ne pas utiliser de fichier, et d'employer les techniques présentées dans la suite de cette section.

Reviewer 2: ```suggestion
Pour éviter cela, il est possible d'adosser un fichier sur le disque dur à la base de données en mémoire au moment de la création de la connexion. Cela ralentit considérablement les calculs, et ne permet pas toujours d'obtenir un résultat. Il est recommandé de ne pas utiliser de fichier, et d'employer plutôt les techniques présentées dans la suite de cette section.
```

- Avec l'évaluation différée, l'ensemble de la requête est exécutée à chaque appel à `collect()`. Il peut être utile de conserver des résultats pour ne pas refaire les mêmes calculs.
- Certaines requêtes sont trop compliquées pour le moteur SQL et/ou pour la traduction automatique. Par contre, `arrow::write_dataset()` sait faire les calculs par morceaux automatiquement, et libère la mémoire au fur et à mesure. Lorsqu'un calcul ne passe pas, on peut tenter de le découper en passant par une base intermédiaire :
```{r eval=FALSE}
con %>% calcul1() %>% 
  arrow::to_arrow() %>% arrow::write_dataset("base_intermediaire")
arrow::open_dataset("base_intermediaire") %>% arrow::to_duckdb(con) %>% 
  calcul2()
```

**Partitionner les données**.

- Pour exécuter une fonction fenêtre, il faut pouvoir localiser les données en mémoire avec un _index_.
- Les fichiers `parquet` ont un index `min-max` : les fichiers sont structurés en blocs, et on indique le minimum et maximum des valeurs du bloc dans les métadonnées. Ceci permet de sauter la lecture d'un bloc si l'on s'intéresse à des valeurs en dehors de la plage `min-max`, parce que l'on filtre les données par exemple.

> - En SQL, on peut créer un index, mais il faut que les données soient en mémoire, ce qui est incompatible avec des grosses volumétries.

Reviewer 2: ```suggestion
- En SQL, on peut créer un index, mais il faut que les données soient en mémoire, ce qui peut s'avérer être incompatible avec de très grosses volumétries.
```

### Optimisations

Les opérations difficiles en SQL, longues, nécessitant beaucoup de mémoire, sont les fonctions "fenêtre" : jointures, `GROUP BY` avec beaucoup de petits groupes, dédoublonnage, etc. On propose ici quelques techniques pour faire passer ces calculs difficiles.

**Utilisation de la mémoire vive**

Comme expliqué plus haut, les objets manipulés dans cette fiche sont des requêtes SQL, et ne nécessitent pas de mémoire vive. Les données déclarées par `read_parquet` sont stockées sur le disque dur, lues à la demande, et "oubliées" à la fin du calcul. On retourne le _résultat_ du calcul.

Pour les opérations compliquées, il peut être nécessaire de charger les données en mémoire pour effectuer le calcul. En cas de débordement, `duckdb` renvoit un message du type :
```
Error: rapi_execute: Failed to run query
Error: Out of Memory Error: could not allocate block of size 262KB (99.7MB/100.0MB used)
Database is launched in in-memory mode and no temporary directory is specified.
Unused blocks cannot be offloaded to disk.

Launch the database with a persistent storage back-end
Or set PRAGMA temp_directory='/path/to/tmp.tmp'
```

> Pour éviter cela, il est possible d'adosser un fichier sur le disque dur à la base de données en mémoire au moment de la création de la connexion. Cela ralentit considérablement les calculs, et ne permet pas toujours d'obtenir un résultat. Il est recommandé de ne pas utiliser de fichier, et d'employer plutôt les techniques présentées dans la suite de cette section.

Reviewer 4: je pense que ce commentaire est à modérer : en gros oui l'utilisation d'un fichier peut prendre beaucoup de temps mais duckdb fait du débordement uniquement s'il en a besoin. Par exemple, s'il n'a que 20Go et qu'il arrive en limite, il va commencer à écrire disque. Du coup s'il a besoin de 21Go il va écrire +/- 1Go. Ca peut éviter de revoir son traitement pour le splitter
